posts/spring2021/js-chats-6-hooks/index.md

```diff
@@ -407,31 +406,33 @@ Yay! It works!
+Generally, the usage for `useReducer()` could be summarized to the following three situations
+
+- You want to use `useReducer()` when there is complex state logic.
+- You want to use `useReducer()` for state management.
+- You want to use `useReducer()` when you want to think of the actions in an atomic way.
 
 ## useContext
 
 ### Problem w/o state management: Props Drilling
 
-![props drilling](images/props-drilling.png)
-
-Look at the screenshot above, 
+![props drilling](images/props-drilling1.png)
 
-### Flux Architecture
+Look at the screenshot above, when the concept of state management hasn't been introduced, if you want to pass a prop message from the `GreatGrandParent` to the deep child `Display` component for showing the message, you need to pass this prop through all the ancestors of `Display` in order to do that. This is the problem of props drilling, where you have to pass a prop through several components in order to use that prop. Even though this still looks fine, it'll become more complicated when many components require props from `GreatGrandParent` and it could quickly become a mess.
 
-In previous discussion, we mentioned reducer is also a concept in Redux. In fact, Redux is an implementation of Facebook's flux architecture.
-
-Imagine you have a project with multiple React components. For these components, they sometimes need to interact with each other by passing props and states. Without redux, you can only have a top-down data passing from root component to each of the component that requires a parent state, and the entire graph forms a directed acyclic graph. However, with the introduction of the flux architecture, all the components' states can be centralized to a structure called "store":
+The problem without an ultimate state manager looks like the following graph. In order to let child components to use the state from the root component (assume the blue circle is the root component), we need to perform a DFS-like graph traversal in order to let the state be used. This is inefficient in itself, but what makes matters worse is that if you store every state in the root component if one state has changed, every component needs to be re-rendered since every node is essentially a child of the root component, this is very inefficient.
```

Comment:
Can we change the start sentence to "The problem without an ultimate state manager can be illustrated by the following graph."

Reply:
done

Reply:
And the following sentence should be: "let child components use the state from the root component"(without "to"). I should really create a counter for my grammar mistakes.